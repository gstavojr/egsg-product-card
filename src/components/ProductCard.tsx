import React, { createContext, CSSProperties } from 'react';

import { useProduct } from '../hooks/useProduct';
import {
  iProductContextProps,
  iProduct,
  iOnChangeArgs,
  iInitialValues,
} from '../interfaces/product.interfaces';

import styles from '../styles/styles.module.css';
import { iProductCardHandles } from '../interfaces/product.interfaces';

// =======================================================================================
// Context
// =======================================================================================
export const ProductContext = createContext({} as iProductContextProps);
const { Provider } = ProductContext;

export interface iProps {
  product: iProduct;
  // children?: ReactElement | ReactElement[];
  children?: (arg: iProductCardHandles) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: iOnChangeArgs) => void;
  value?: number;
  initialValues?: iInitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: iProps) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider
      value={{
        counter,
        product,
        maxCount,
        increaseBy,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children &&
          children({
            count: counter,
            isMaxCountReached,
            maxCount: initialValues?.maxCount,
            product,

            increaseBy,
            reset,
          })}
      </div>
    </Provider>
  );
};
