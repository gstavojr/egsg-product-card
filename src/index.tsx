// Exportaciones que seran vista desde el mundo exterior

export * from './components';
