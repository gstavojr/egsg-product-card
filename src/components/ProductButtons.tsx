import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';

// Estilos mediante modulos
import styles from '../styles/styles.module.css';

export interface iProps {
  className?: string;
  style?: CSSProperties;
}

// Mini componente de botones
export const ProductButtons = ({ className, style }: iProps) => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  // si counter === maxCount desactivar el boton.
  const isMaxReached = useCallback(
    // verficar que el maxCount exista, si no existe es false -> !!maxCount
    // si existe procese a evaluar.
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  );

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        {' '}
        -{' '}
      </button>

      <div className={styles.countLabel}>{counter}</div>

      <button
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disable}`}
        onClick={() => increaseBy(1)}
      >
        {' '}
        +{' '}
      </button>
    </div>
  );
};
