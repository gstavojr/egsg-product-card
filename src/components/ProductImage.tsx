import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

// Estilos mediante modulos
import noImg from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface iProps {
  className?: string;
  img?: string;
  style?: CSSProperties;
}

// Mini componente de una imagen
export const ProductImage = ({ img = '', className, style }: iProps) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = product.img;
  } else {
    imgToShow = noImg;
  }

  return (
    <img
      src={imgToShow}
      className={`${styles.productImg} ${className}`}
      alt="Product Imagen"
      style={style}
    />
  );
};
