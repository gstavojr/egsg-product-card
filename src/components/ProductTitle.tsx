import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';

// Estilos mediante modulos
import styles from '../styles/styles.module.css';

export interface iProps {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

// Mini componente de una titulo
export const ProductTitle = ({ title, className, style }: iProps) => {
  const { product } = useContext(ProductContext);
  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {title ? title : product.title}
    </span>
  );
};
