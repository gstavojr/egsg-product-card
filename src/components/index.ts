import { ProductCard as ProductCardHOC } from './ProductCard';
import { iProductCardHOCProps } from '../interfaces/product.interfaces';

import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export { ProductButtons } from './ProductButtons';
export { ProductTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';

// Asignar nuevas propiedades al productCard
export const ProductCard: iProductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});

export default ProductCard;
