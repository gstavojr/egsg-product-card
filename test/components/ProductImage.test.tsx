import React from 'react';
import render from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('Debe mostrar el componente correctamente con la imagen personalizada', () => {
    const wrapper = render.create(<ProductImage img="hola mundo" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostrar el componente con la imagen del producto', () => {
    const wrapper = render.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
