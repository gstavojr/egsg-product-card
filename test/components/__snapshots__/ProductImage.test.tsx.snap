// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="Product Imagen"
  className="productImg undefined"
  src="hola mundo"
/>
`;

exports[`ProductImage debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product Imagen"
    className="productImg undefined"
    src="./coffe-mug.png"
  />
</div>
`;
